Lazy-load images in the artwork grid

zephraph opened the pull request:
This adds lazy loading to the artwork grid. Addresses [PLATFORM-1164](https://artsyproduct.atlassian.net/browse/PLATFORM-1164). This demo uses 2G speed.

![2019-02-25 18-48-47 2019-02-25 18_52_45](https://user-images.githubusercontent.com/3087225/53377054-92760e00-392e-11e9-8852-4726af236c77.gif)

The placeholder slightly pulses to give the impression of something happening (suggested by Nicole) and fades in at our standard `250ms`. 

### Todo
- [x] Add the ability to specify how many images should be eagerly loaded for each grid
- [x] Setup up appropriate eager loading behaviors for artist, artwork, collect, collection

artsyit commented:
Deploy preview for *artsy-reaction* ready!

Built with commit 42b3aa764dad6d09d2194b1e9915bb8bec987f74

https://deploy-preview-2064--artsy-reaction.netlify.com

damassi commented on src/Components/Artwork/GridItem.tsx:
```diff
@@ -1,27 +1,44 @@
+    ${pulse} 2s ease-in-out infinite;
+  `
+
+const Image = styled(LazyLoadImage)`
```
Wondering, should this be configured so that it lazy loads below the fold content? Similar to how we trigger the "Recently Viewed" rail to appear before the user has scrolled to it. Seeing the image jump in in real time isn't too optimial.

damassi replied:
(Ah, I see the `todo` item in the description)

zephraph replied:
The library offers a threshold that we can use to set the load distance. It's currently 100px below the fold. Fast scrollers are one case, slow phones are another. We can tweak the setting if needed. Might be worth standardizing on a distance below-the-fold loading threshold.

damassi replied:
Cool, I bet a couple hundred pixels would be sufficient.

damassi commented on src/Components/Artwork/GridItem.tsx:
```diff
@@ -34,6 +51,7 @@ interface Props extends React.HTMLProps<ArtworkGridItemContainer> {
 
 interface State {
   isMounted: boolean
+  isImageLoaded: boolean
```
Should the `<ArtworkGrid>` take a `lazyLoad` configuration prop? I'm thinking of pages like: https://www.artsy.net/collect where the the grid is at the top of the screen and you'd want to be able to see the images on load and with JS disabled.

zephraph replied:
I plan to have a prop for `ArtworkGrid` that accepts a number of images to eagerly load. So for the collect page 9 images on desktop, 6 images on mobile would be ideal. For artist/artwork page 3 to 6 images makes sense to eagerly load on desktop, but you'd probably want none on mobile. 

As per the js disabled bit... that should generally be handled by the lib. I'll look into if it is, if not, I'll PR it.

alloy replied:
Is there no way for the lib to just eager load above the fold? (Not sure if the above thread is about that.)

zephraph replied:
Nope. I mean, when you load the page and the js initializes it will auto load anything in view (and within the threshold). But the above the fold loading decision has to be made at SSR which falls back to the traditional not knowing the user’s browser size.

zephraph replied:
I've manually configured the lazy-loading values for each case. For artwork grids completely above the fold (think collect, works-for-you) there are 9 preloaded images. That seems to cover a pretty tall screen.

alloy commented on src/Components/Artwork/GridItem.tsx:
```diff
@@ -167,7 +193,11 @@ class ArtworkGridItemContainer extends React.Component<Props, State> {
+              style={{ opacity: isImageLoaded ? "1" : "0" }}
+              src={this.getImageUrl()}
+              onLoad={this.imageLoaded.bind(this)}
+            />
```
Will there be a `img` tag with `src` attribute coming from the SSR content? If not, does that affect SEO in any way?

cavvia replied:
I had the same question actually... will point nicolette to this for her thoughts.

zephraph replied:
From an SSR perspective, no. Based on google’s own recommendations, this should be fine as long as the image does indeed pull in as the bot scrolls. https://developers.google.com/search/docs/guides/lazy-loading. Traditionally lazy-loaded imgs use the `data-src` attribute as a placeholder before loading, but that’s not the case with this library.

NicoletteChao replied:
Would just raise that traditionally `src` and `alt` are used to help Google understand what the image is. Since `src` is typically a cloudfront URL for us without a description of the artwork I think it's okay to forego that, however, can we add `alt` text to the images? They could just be the artwork title. 

I believe we used to have these on our artist pages, or at least, in the snapshot we provided to Google through reflection.

zephraph replied:
I actually noticed we're not really doing a good job of that anymore... I'll try to fix it up. 👍

zephraph replied:
I've buttoned this up. `title` attributes come from the artwork `title` field and the `alt` attribute is provided by the `image_title` field (that has details such as author and year).

zephraph commented:
I made some general updates. I don't use the lazy-load component on preloaded images. That helps to not trigger the lazy loading animation unnecessarily. I also added some `noscript` tags to cover the case where js doesn't load. I added `title` and `alt` tags to the images. I think the biggest thing remaining is to tweak the amount of preloaded images on certain pages (like the Collect page).

damassi commented on src/Apps/Artwork/Components/OtherWorks/ArtworkContexts/ArtworkGrids/ArtistArtworkGrid.tsx:
```diff
@@ -45,6 +45,7 @@ export class ArtistArtworkGrid extends React.Component<ArtistArtworkGridProps> {
         <ArtworkGrid
           artworks={artist.artworks_connection}
           columnCount={[2, 3, 4]}
+          preloadImageCount={0}
```
Should this be a default prop?

zephraph replied:
This definitely shouldn't be! That'd mean they always lazy load by default... which, I mean might be good sometimes, but might not others.

damassi commented on src/Apps/Collect/Components/Base/CollectArtworkGrid.tsx:
```diff
@@ -86,6 +86,7 @@ class CollectArtworkGrid extends Component<Props, LoadingAreaState> {
                     <ArtworkGrid
                       artworks={artworks as any}
                       columnCount={columnCount}
+                      preloadImageCount={9}
```
Or should _this_ be a default prop 😄

zephraph replied:
It _could_ be. Right now I have the default as 6 which seems to cover most cases.

damassi commented on src/Components/Artwork/GridItem.tsx:
```diff
@@ -73,6 +101,12 @@ class ArtworkGridItemContainer extends React.Component<Props, State> {
     }
   }
 
+  imageLoaded() {
```
`imageLoaded = () =>` and you can loose the `.bind` down below.

damassi commented on src/Components/Artwork/GridItem.tsx:
```diff
@@ -167,7 +202,28 @@ class ArtworkGridItemContainer extends React.Component<Props, State> {
+                alt={artwork.image_title}
+                style={{ opacity: isImageLoaded ? "1" : "0" }}
+                src={this.getImageUrl()}
+                onLoad={this.imageLoaded.bind(this)}
```
See above

damassi commented on src/Components/Artwork/GridItem.tsx:
```diff
@@ -167,7 +202,28 @@ class ArtworkGridItemContainer extends React.Component<Props, State> {
+                alt={artwork.image_title}
+                src={this.getImageUrl()}
+              />
+            </noscript>
```
SEO?

zephraph replied:
Neither. Just the no js use case. I mean, it _might_ be unnecessary... so much of our site wouldn't work without js anyway. But.. yeah. SEO effectively ignores this and I doubt it'd be used for accessibility... though I'm not 100% sure of that.

damassi replied:
OH duh, for no JS -- but that said, our new pages are totally browsable / usable without JS turned on so we want to keep things fluid where possible -- this seems like a good solution 👍 .

damassi commented:
Pulled this down and scrolled and the fade-in distance timings felt very natural / expected. Turned off JS and also looked great 💯

Mergin

artsyit commented:
<!-- GITHUB_RELEASE COMMENT: released -->
:rocket: PR was released in v15.2.0 :rocket:
